consensus/istanbul/ibft/core/core.go

```diff
@@ -332,11 +332,37 @@ func (c *core) newRoundChangeTimer() {
 	c.stopTimer()
 
 	// set timeout based on the round number
-	timeout := time.Duration(c.config.GetConfig(c.current.Sequence()).RequestTimeout) * time.Millisecond
+	baseTimeout := time.Duration(c.config.GetConfig(c.current.Sequence()).RequestTimeout) * time.Millisecond
 	round := c.current.Round().Uint64()
-	if round > 0 {
-		timeout += time.Duration(math.Pow(2, float64(round))) * time.Second
+	maxRequestTimeout := time.Duration(c.config.GetConfig(c.current.Sequence()).MaxRequestTimeoutSeconds) * time.Second
+
+	// If the upper limit of the request timeout is capped by small maxRequestTimeout, round can be a quite large number,
+	// which leads to float64 overflow, making its value negative or zero forever after some point.
+	// In this case we cannot simply use math.Pow and have to implement a safeguard on our own, at the cost of performance (which is not important in this case).
+	var timeout time.Duration
+	if maxRequestTimeout > time.Duration(0) {
+		timeout = baseTimeout
+		for i := uint64(0); i < round; i++ {
+			timeout = timeout * 2
+			if timeout > maxRequestTimeout {
+				timeout = maxRequestTimeout
+				break
+			}
+		}
+		// prevent log storm when unexpected overflow happens
+		if timeout < baseTimeout {
+			c.currentLogger(true).Error("IBFT: Possible request timeout overflow detected, setting timeout value to maxRequestTimeout",
+				"timeout", timeout.Seconds(),
+				"max_request_timeout", maxRequestTimeout.Seconds(),
+			)
+			timeout = maxRequestTimeout
+		}
+	} else {
+		// effectively impossible to observe overflow happen when maxRequestTimeout is disabled
+		timeout = baseTimeout * time.Duration(math.Pow(2, float64(round)))
 	}
```

Comment:
Was this cherry picked from Consensys/quorum? If the concern is potential overflow of `math.Pow(2, float64(round))`, instead figure out the max round allowed via `Log2()`, consider the following suggested pseudo-code:

```suggestion
      timeout := baseTimeout 
      maxRound := round
      if maxRequestTimeout > time.Duration(0) {
        maxRound = uint64(math.Floor(math.Log2(float64(maxRequestTimeout / baseTimeout))))
      }
      if round > maxRound {
        timeout = maxRequestTimeout
      }else{
        timeout = baseTimeout * time.Duration(math.Pow(2, float64(round)))
      }
```

Reply:
i think this is a better approach, using `log2` for faster computation

Since `log2` can get create a binary representation, then count number of bits (length) 
https://stackoverflow.com/questions/47074126/log2-of-an-integer-that-is-a-power-of-2